src/Utilities/MDFlatten.jl

```diff
@@ -0,0 +1,98 @@
+"""
+Provides the [`mdflatten`](@ref) function that can "flatten" Markdown objects into
+a string, with formatting etc. stripped.
+
+Note that the tests in `test/mdflatten.jl` should be considered to be the spec
+for the output (number of newlines, indents, formatting, etc.).
+"""
+module MDFlatten
+
+export mdflatten
+
+import Base.Markdown:
+    MD, BlockQuote, Bold, Code, Header, HorizontalRule,
+    Image, Italic, LaTeX, LineBreak, Link, List, Paragraph, Table
+
+"""
+Convert a Markdown object to a `String` of only text (i.e. not formatting info).
+
+It drop most of the extra information (e.g. language of a code block, URLs)
+and formatting (e.g. emphasis, headers). This "flattened" representation can
+then be used as input for search engines.
+"""
+function mdflatten(md)
+    io = IOBuffer()
+    mdflatten(io, md)
+    takebuf_string(io)
+end
+
+mdflatten(io, md) = mdflatten(io, md, md)
+mdflatten(io, md::MD, parent) = mdflatten(io, md.content, md)
+mdflatten(io, vec::Vector, parent) = map(x -> mdflatten(io, x, parent), vec)
+function mdflatten(io, vec::Vector, parent::MD)
+    # this special case separates top level blocks with newlines
+    for md in vec
+        mdflatten(io, md, parent)
+        print(io, "\n\n")
+    end
+end
+
+# Block level MD nodes
+mdflatten{N}(io, h::Header{N}, parent) = mdflatten(io, h.text, h)
+mdflatten(io, p::Paragraph, parent) = mdflatten(io, p.content, p)
+mdflatten(io, bq::BlockQuote, parent) = mdflatten(io, bq.content, bq)
+mdflatten(io, ::HorizontalRule, parent) = nothing
+function mdflatten(io, list::List, parent)
+    for (idx, li) in enumerate(list.items)
+        for (jdx, x) in enumerate(li)
+            mdflatten(io, x, list)
+            jdx == length(li) || print(io, '\n')
+        end
+        idx == length(list.items) || print(io, '\n')
+    end
+end
+function mdflatten(io, t::Table, parent)
+    for (idx, row) = enumerate(t.rows)
+        for (jdx, x) in enumerate(row)
+            mdflatten(io, x, t)
+            jdx == length(row) || print(io, ' ')
+        end
+        idx == length(t.rows) || print(io, '\n')
+    end
+end
+
+# Inline nodes
+mdflatten(io, text::AbstractString, parent) = print(io, text)
+mdflatten(io, link::Link, parent) = mdflatten(io, link.text, link)
+mdflatten(io, b::Bold, parent) = mdflatten(io, b.text, b)
+mdflatten(io, i::Italic, parent) = mdflatten(io, i.text, i)
+mdflatten(io, i::Image, parent) = print(io, "(Image: $(i.alt))")
+mdflatten(io, m::LaTeX, parent) = print(io, "(equation)")
```

Comment:
For page titles such as the "LaTeX syntax" one, it's printed as "(equation) syntax". Perhaps just strip punctuation characters, but leave the rest intact? Something like `replace(m.formula, r"[^\w\s]", "")` might be enough.

I don't think it's worth spending too much time on though, since page titles shouldn't really contain anything too complex and can be set explicitly in the `pages` keyword if needed.